File: julia/two-bucket/two-bucket.jl
function twobucket(bucket1, bucket2, goal, start)
    if goal % gcd(bucket1, bucket2) != 0 || goal > max(bucket1, bucket2)
        throw(DomainError("Impossible")) 
    end

    steps = 1
    found = nothing
    seen = Set([(0, 0), (bucket1, 0), (0, bucket2)])
    status = start == 1 ? s = Set([(bucket1, 0), ]) : Set([(0, bucket2), ])

    # have we already found a solution?
    for (v1, v2) in status
        if v1 == goal || v2 == goal
            found = (v1, v2)
        end
    end

    # step
    while found == nothing
        prev_status = status
        status = Set()
        steps += 1
        for (vol1, vol2) in prev_status
            # empty/fill one bucket:
            next_steps = Set([(vol1, 0), (0, vol2), (vol1, bucket2), (bucket1, vol2)])

            # transfers:
            transfer_1to2 = min(vol1, bucket2 - vol2)
            push!(next_steps, (vol1 - transfer_1to2, vol2 + transfer_1to2))

            transfer_2to1 = min(vol2, bucket1 - vol1)
            push!(next_steps, (vol1 + transfer_2to1, vol2 - transfer_2to1))

            # update sets
            for n in setdiff(next_steps, seen)
                push!(status, n)
                push!(seen, n)
            end
        end

        # have we found a solution?
        for (v1, v2) in status
            if v1 == goal || v2 == goal
                found = (v1, v2)
                break
            end
        end
    end

    # Finish
    return found[1] == goal ? (steps, 1, found[2]) : (steps, 2, found[1])
end

File: fsharp/two-bucket/TwoBucket.fs
module TwoBucket

type Bucket =
    | One
    | Two
    
type Result = { Moves:int; GoalBucket:Bucket; OtherBucket:int }

let measure bucketOne bucketTwo goal startBucket =
    let isGoal (b1, b2) = b1 = goal || b2 = goal
    
    let isFound stats =
        let matches = Set.filter isGoal stats
        if matches.IsEmpty then None else Some (Set.toArray matches)
    
    let branch (b1, b2) =
        let transfer1to2 = min b1 (bucketTwo - b2)
        let transfer2to1 = min b2 (bucketOne - b1)
        [ (b1, 0); (0, b2); (b1, bucketTwo); (bucketOne, b2); // empty/fill one bucket
          (b1 - transfer1to2, b2 + transfer1to2) // transfer 1 -> 2
          (b1 + transfer2to1, b2 - transfer2to1) ] // transfer 2 -> 1

    let rec step steps prevSeen prevStatus =
        let possibleSteps = Set (List.collect branch (Set.toList prevStatus)) - prevSeen
        match isFound possibleSteps with
        | Some m ->
            let (b1, b2) = m[0]
            if b1 = goal
                then { Moves = steps + 1; GoalBucket = Bucket.One; OtherBucket = b2 }
                else { Moves = steps + 1; GoalBucket = Bucket.Two; OtherBucket = b1 }
        | None -> step (steps + 1) (prevSeen + possibleSteps) (prevStatus + possibleSteps)
            
    let seen = Set [(0, 0); (bucketOne, 0); (0, bucketTwo)]
    let status = if startBucket = Bucket.One then Set [ (bucketOne, 0); ] else Set [ (0, bucketTwo); ]
    match isFound status with
    | Some m ->
        let (b1, b2) = m[0]
        if b1 = goal
            then { Moves = 1; GoalBucket = Bucket.One; OtherBucket = b2 }
            else { Moves = 1; GoalBucket = Bucket.Two; OtherBucket = b1}
    | None -> step 1 seen (Set status)
